### PR TITLE
Implement automations using Marionette Protocol (closes #2623)

### LINKED ISSUE #2623: Implement user action emulation using Firefox Marionette protocol

### PR DESCRIPTION
**Pull request by AndreyBelym**

### CODE REVIEW

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

File: src/browser/provider/built-in/firefox/marionette-client.js [hunk 1]
```diff
@@ -192,8 +274,124 @@ module.exports = class MarionetteClient {
...
     async quit () {
         await this._getResponse({ command: 'quit' });
     }
+
+    async cancelAction () {
+        this.events.emit('cancel');
+    }
+
+    async executeCommand (msg) {
+        if (this.executingAction)
```
Comment (miherlosev) [inline src/browser/provider/built-in/firefox/marionette-client.js]: It's a very large method. I propose to restructure it as follows:
```js
async executeCommand (msg) {
   if (msg.modifiers && !msg.clearMods) 
       this._pefromKeyboradAction (...);

  swith (msg.type) {
     'move': this._performMouseMoveAction (....);
      break;
     'click':
     'right-click':
      ....
      this.performMouseAction (...);
      break;
      ...
   
}

}
```

**Comment by AndreyBelym** [comment]
@testcafe-build-bot retest

File: src/client/automation/playback/click/index.js [hunk 2]
```diff
@@ -6,22 +6,27 @@ import { focusAndSetSelection, focusByRelatedElement } from '../../utils/utils';
...
-var domUtils   = testCafeCore.domUtils;
-var styleUtils = testCafeCore.styleUtils;
-var eventUtils = testCafeCore.eventUtils;
-var arrayUtils = testCafeCore.arrayUtils;
-var delay      = testCafeCore.delay;
+const domUtils   = testCafeCore.domUtils;
+const styleUtils = testCafeCore.styleUtils;
+const eventUtils = testCafeCore.eventUtils;
+const arrayUtils = testCafeCore.arrayUtils;
+const marionetteUtils = testCafeCore.marionetteUtils;
```
Comment (miherlosev) [inline src/client/automation/playback/click/index.js]: I guess that `marionetteClient` or `marionetteDriver` name is a more suitable.

File: src/test-run/index.js [hunk 3]
```diff
@@ -53,6 +53,21 @@ const MAX_RESPONSE_DELAY              = 2 * 60 * 1000;
 
 const ALL_DRIVER_TASKS_ADDED_TO_QUEUE_EVENT = 'all-driver-tasks-added-to-queue';
 
+function getMarionetteClient (connection) {
```
Comment (miherlosev) [inline src/test-run/index.js]: I propose to move this function to the `\src\browser\provider\built-in\firefox\get-marionette-client.js` file because it is not related to the `TestRun` class.

File: src/browser/provider/built-in/firefox/marionette-client.js [hunk 4]
```diff
@@ -72,15 +101,58 @@ module.exports = class MarionetteClient {
...
+            returnPromiseControl = { resolve, reject };
+        });
+
+        chainPromise = chainPromise
+            .then(() => {
+                const cancelPromise = promisifyEvent(this.events, 'cancel');
+
+                cancelPromise.then(() => {
+
+                    returnPromiseControl.resolve = () => {};
```
Comment (miherlosev) [inline src/browser/provider/built-in/firefox/marionette-client.js]: Use `lodash.noop` function:
```js
returnPromiseControl.resolve = noop;
returnPromiseControl.reject   = noop;
```

File: src/test-run/client-messages.js [hunk 5]
```diff
@@ -1,5 +1,6 @@
 export default {
     ready:                       'ready',
     readyForBrowserManipulation: 'ready-for-browser-manipulation',
-    waitForFileDownload:         'wait-for-file-download'
+    waitForFileDownload:         'wait-for-file-download',
+    performActions:              'perform-actions'
```
Comment (miherlosev) [inline src/test-run/client-messages.js]: I propose to rename this to more specialized - `perform-marionette-action`.

File: src/client/test-run/iframe.js.mustache [hunk 6]
```diff
@@ -1,10 +1,10 @@
 (function () {
     var IframeDriver = window['%testCafeIframeDriver%'];
     var driver       = new IframeDriver({{{testRunId}}}, {
-        selectorTimeout: {{{selectorTimeout}}},
-        pageLoadTimeout: {{{pageLoadTimeout}}},
-        dialogHandler:   {{{dialogHandler}}},
-        speed:           {{{speed}}}
+        selectorTimeout:    {{{selectorTimeout}}},
```
Comment (miherlosev) [inline src/client/test-run/iframe.js.mustache]: You added additional indents here and in `src/client/test-run/index.js.mustache`

File: src/test-run/index.js [hunk 7]
```diff
@@ -639,6 +657,11 @@ export default class TestRun extends EventEmitter {
 
         return await getLocation();
     }
+
+    async [CLIENT_MESSAGES.performActions] (msg) {
+        await this.marionetteClient.executeCommand(msg);
```
Comment (miherlosev) [inline src/test-run/index.js]: Add logging:
```js
this.debugLog.driverMessage(msg);
```

File: src/client/core/utils/marionnette.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+export default {
+    enabled: false,
+
+    performAction (args) {
+        disableRealEventsPreventing();
+
+        const msgArgs = { disableResending: true, cmd: MESSAGE.performActions };
+
+        nativeMethods.objectKeys.call(window.Object, args).forEach(prop => {
```
Comment (miherlosev) [inline src/client/core/utils/marionnette.js]: Rewrite using `hammerhead.util.extend` method

File: src/browser/provider/built-in/firefox/mime-types.js [hunk 9]
```diff
@@ -0,0 +1 @@
...-uuencode;text/x-vcalendar;text/x-vcard;application/x-cdlink;application/vnd.vsf;model/vrml;application/vnd.vcx;model/vnd.mts;model/vnd.vtu;application/vnd.visionary;video/vnd.vivo;applicatin/ccxml+xml,;application/voicexml+xml;application/x-wais-source;application/vnd.wap.wbxml;image/vnd.wap.wbmp;audio/x-wav;application/davmount+xml;application/x-font-woff;application/wspolicy+xml;image/webp;application/vnd.webturb;application/widget;application/winhlp;text/vnd.wap.wml;text/vnd.wap.wmlscript;application/vnd.wap.wmlscriptc;application/vnd.wordperfect;application/vnd.wt.stf;application/wsdl+xml;image/x-xbitmap;image/x-xpixmap;image/x-xwindowump;application/x-x509-ca-cert;application/x-xfig;application/xhtml+xml;application/xml;application/xcap-diff+xml;application/xenc+xml;application/patch-ops-error+xml;application/resource-lists+xml;application/rls-services+xml;aplication/resource-lists-diff+xml;application/xslt+xml;application/xop+xml;application/x-xpinstall;application/xspf+xml;application/vnd.mozilla.xul+xml;chemical/x-xyz;text/yaml;application/yang;application/yin+xml;application/vnd.ul;application/zip;application/vnd.handheld-entertainment+xml;application/vnd.zzazz.deck+xml';
```
Comment (miherlosev) [inline src/browser/provider/built-in/firefox/mime-types.js]: Remove this file

File: package.json [hunk 10]
```diff
@@ -49,7 +49,7 @@
     "await"
   ],
   "scripts": {
-    "test": "gulp travis",
+    "test": "gulp travis --dev",
```
Comment (miherlosev) [inline package.json]: Revert this change

File: src/client/automation/playback/click/index.js [hunk 11]
```diff
@@ -69,6 +80,28 @@ export default class ClickAutomation extends VisibleElementAutomation {
         eventUtils.bind(element, 'click', onclick);
     }
 
+    _bindMouseUpListener () {
```
Comment (miherlosev) [inline src/client/automation/playback/click/index.js]: `_bindMouseUpAndClickListeners`

**Comment by AndreyBelym** [comment]
1. Agreed, but it won't be a client but functional test. because we need to run both client and server scripts to test this.

2. Agreed, It's #2361, and I've told about it at the last sprint planning.

**Comment by AlexKamaev** [comment]
I'll close this PR since it is not under active development.
